rework `Chip` and `ChipContainer`

joe-s-avaya opened the pull request:
[Link to updated Chip Stories](https://deploy-preview-60--neo-react-library-storybook.netlify.app/?path=/story/components-chips)

**Before tagging the team for a review, I have done the following:**
- [x] reviewed my code changes
- [x] ensured that all tests pass
- [x] updated the link at the top of this PR (or remove it if not applicable)
- [x] added any important information to this PR (description, images, GIFs, ect.)
- [x] tagged Matt if any visual changes have occurred
- [x] done an accessibility check on my work

Nearly all files have been renamed, and the only one that wasn't that needs to be paid close attention to is [Chips.tsx](https://github.com/avaya-dux/neo-react-library/blob/b95264c99b7422d1caf0de54c532981a06e1bea9/src/components/Chip/Chip.tsx), which I recommend not looking at the diff for, but looking at directly.

Unit test code coverage is 100% for the Chip and Chip Container.

Chip's are no longer buttons as that breaks a11y. But they do accept a `onClose` prop for when the user makes the chip closable (the close icon is a button, which also has an aria label).

Note that I have _not_ done the "[Expandable Chip](https://design.avayacloud.com/components/web/chip-web#ExpandableChips)" as I checked with Matt and he is unsure if we need to support that option. If Matt comes back and tells me that we do need it I'll add it.

**GIF Example**:
![2022-11-03 08 18 22](https://user-images.githubusercontent.com/55896546/199761386-fb337f8a-0050-45e3-965f-7a8614594d04.gif)

matthewAvaya commented:
Default
- Add 8px padding between each chip example (for all examples)

Closable
- Do we need a chip for 'Not Closable'? Seems odd that it's under the Closable category. 

Chips Container Examples
- Avatar looks big. On the design site, the avatar is scaled at 28x28px.
- I believe the title for the section with the avatars should be 'Chips with Avatar' like it is on the design portal. 

Templated
- What's the difference between this chip compared to the chip under the Default category? They look identical. 

Has the accessibility part of this been done? The screen reader doesn't work and the focus order seems incomplete.

yangAtSpoken reviewed (commented):
The changes lost some of the nuances:
* icon chip supports rtl; closable does not;
* when chips are inside chip container, they need to have neo-chips__item class.
* avatar needs to have size=sm attribute.
* Tooltip is gone.

joe-s-avaya commented:
> Default
> 
> * Add 8px padding between each chip example (for all examples)
> 
> Closable
> 
> * Do we need a chip for 'Not Closable'? Seems odd that it's under the Closable category.
> 
> Chips Container Examples
> 
> * Avatar looks big. On the design site, the avatar is scaled at 28x28px.
> * I believe the title for the section with the avatars should be 'Chips with Avatar' like it is on the design portal.
> 
> Templated
> 
> * What's the difference between this chip compared to the chip under the Default category? They look identical.
> 
> Has the accessibility part of this been done? The screen reader doesn't work and the focus order seems incomplete.


Will add padding between chips when in a container.

In the "Closable" section, I added the "Not Closable" chip just so that the user can easily see the difference between the two. Especially when tabbing. 

Avatar is fixed. Title of that section is also fixed.

Templated is there for playing with the different options for a chip. I've added a GIF example in the description.

A11y has been done. I'll ping you about what exactly you're seeing.

joe-s-avaya commented:
> The changes lost some of the nuances:
> 
> * icon chip supports rtl; closable does not;
> * when chips are inside chip container, they need to have neo-chips__item class.
> * avatar needs to have size=sm attribute.
> * Tooltip is gone.

RTL: I see no reason why that would be. So I'm reluctant to make that change unless I can find a compelling reason to add it.
`neo-chips__item`: will add, I missed that.
Avatar size: oh yeah, good call, updated.
Tooltip: there's no reason to have props for tooltip. The dev can just wrap the chip (or it's content) with a tooltip.

yangAtSpoken commented:
> > The changes lost some of the nuances:
> > 
> > * icon chip supports rtl; closable does not;
> > * when chips are inside chip container, they need to have neo-chips__item class.
> > * avatar needs to have size=sm attribute.
> > * Tooltip is gone.
> 
> RTL: I see no reason why that would be. So I'm reluctant to make that change unless I can find a compelling reason to add it. `neo-chips__item`: will add, I missed that. Avatar size: oh yeah, good call, updated. Tooltip: there's no reason to have props for tooltip. The dev can just wrap the chip (or it's content) with a tooltip.

RTL requirements are from https://design.avayacloud.com/components/web/chip-web 
icon can be placed on either left or right;
closeable chip always have x on the right;
@matthewAvaya can you confirm this?

enrique-prado commented on src/components/Chip/Chip.tsx:
```diff
@@ -1,33 +1,70 @@
+  disabled?: boolean;
+  icon?: IconNamesType;
+  onClose?: MouseEventHandler<HTMLButtonElement>;
+  variant?: "alert" | "default" | "info" | "success" | "warning";
```
I thought Matt said chips don't need variants at Wednesday's meeting. Has this changed?

joe-s-avaya replied:
That was the dropdown/expandable chip, which he is still gathering information about. So yes, we _do_ want variants still.

enrique-prado replied:
ok, sounds good

yangAtSpoken commented:
Another requirement is that Avatar chip has does not allow icon.

yangAtSpoken commented on src/components/Chip/ChipsContainer.tsx:
```diff
@@ -0,0 +1,30 @@
+    {Children.map(children, (child, index) =>
+      cloneElement(child, {
+        key: index,
+        className: "neo-chips__item",
```
Does this add to "className" or replace it?

joe-s-avaya replied:
Whoops, replaces it. Will update, good catch, thanks!

joe-s-avaya replied:
Updated. I also noticed that I was extending the HTML Div Element, but I wasn't actually applying any of those props. Thus, I've also added `{...rest}` to both `Chip` and `ChipsContainer`.

yangAtSpoken commented on src/components/Chip/Chip.tsx:
```diff
@@ -1,33 +1,70 @@
-  onClick?: React.MouseEventHandler;
-  variant?: Variants;
-};
+  closeButtonAriaLabel?: string;
```
I do not think we have closable avatar chip per https://design.avayacloud.com/components/web/chip-web

joe-s-avaya replied:
Just checked with Matt, you are correct. Updated.

enrique-prado commented on src/components/Chip/index.ts:
```diff
@@ -1,13 +1,2 @@
-export type { ClosableChipProps } from "./ClosableChip";
-export type { ExpandableChipProps } from "./ExpandableChip";
-export type { IconChipProps } from "./IconChip";
-export type { ChipContainerProps } from "./ChipContainer";
```
Did all these files get renamed or deleted? I was expecting the former.

joe-s-avaya replied:
Deleted. Whereas before we have a separate file for every variant, now we have one single file that takes props. Thus I was able to reduce from many chip components to just two (`Chip` and `ChipContainer`).

enrique-prado replied:
oh, I see. Nice!

enrique-prado reviewed (commented):
Overall the code looks good but I'm a little hesitant to approve because of all the deleted files. It's unclear to me what happened to all that code, I was expecting these files to be renamed not deleted. If you can clarify that would be great. thanks.

enrique-prado commented on src/components/Chip/Chip.stories.tsx:
```diff
@@ -0,0 +1,94 @@
+  disabled: false,
+  icon: undefined,
+  variant: "default",
+};
```
I'm noticing we don't have a story where a chip is created dynamically like it typically happens in a real app.... basically you type a word in an input control and after you hit enter it creates a chip.  It would be nice to have one like that.

joe-s-avaya replied:
Those exist in the [Select Creatable](https://deploy-preview-60--neo-react-library-storybook.netlify.app/?path=/story/components-select--creatable) stories. I don't feel that it's necessary to add a story for that, but if you think that it would add a lot to the DX, I'm happy to add it.

enrique-prado replied:
cool. thanks.

joe-s-avaya commented:
> > > The changes lost some of the nuances:
> > > 
> > > * icon chip supports rtl; closable does not;
> > > * when chips are inside chip container, they need to have neo-chips__item class.
> > > * avatar needs to have size=sm attribute.
> > > * Tooltip is gone.
> > 
> > 
> > RTL: I see no reason why that would be. So I'm reluctant to make that change unless I can find a compelling reason to add it. `neo-chips__item`: will add, I missed that. Avatar size: oh yeah, good call, updated. Tooltip: there's no reason to have props for tooltip. The dev can just wrap the chip (or it's content) with a tooltip.
> 
> RTL requirements are from https://design.avayacloud.com/components/web/chip-web icon can be placed on either left or right; closeable chip always have x on the right; @matthewAvaya can you confirm this?

@yangAtSpoken, just discussed with @matthewAvaya. The RTL ish is something that he is going to discuss with the Localization team and then get back to me on. For now he is comfortable leaving as is.

joe-s-avaya commented:
@matthewAvaya, I've made the updates we discussed, please reviews again: [Link to Chip stories](https://deploy-preview-60--neo-react-library-storybook.netlify.app/?path=/story/components-chips)
